### PR TITLE
[d3d9] Wine test fixes 1

### PR DESCRIPTION
**Pull request by DadSchoorse**
More stuff to make more tests not crash.

### CODE REVIEW

File: src/d3d9/d3d9_device.cpp [hunk 1]
```diff
@@ -936,6 +936,9 @@ namespace dxvk {
     const VkImageSubresource dstSubresource = dstTextureInfo->GetSubresourceFromIndex(dstFormatInfo->aspectMask, dst->GetSubresource());
     const VkImageSubresource srcSubresource = srcTextureInfo->GetSubresourceFromIndex(srcFormatInfo->aspectMask, src->GetSubresource());
 
+    if (unlikely(srcSubresource.aspectMask & (VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT) && m_activeBeginScene))
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:939]: Add () around & comparison

File: src/d3d9/d3d9_device.cpp [hunk 2]
```diff
@@ -2361,11 +2361,14 @@ namespace dxvk {
...
-	if (unlikely(!PrimitiveCount))
-	  return S_OK;
-
     D3D9DeviceLock lock = LockDevice();
 
+    if (unlikely(m_state.vertexDecl == nullptr))
+      return D3DERR_INVALIDCALL;
+
+    if (unlikely(!PrimitiveCount))
+      return S_OK;
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:2370]: Why move this into the lock?

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_device.cpp (hunk 2 above)
Comment (doitsujin) [reply src/d3d9/d3d9_device.cpp:2370]: it accesses m_state so it has to be inside the lock

Comment (DadSchoorse) [reply src/d3d9/d3d9_device.cpp:2370]: Yes, what doitsujin said :frog:

Comment (Joshua-Ashton) [reply src/d3d9/d3d9_device.cpp:2370]: Got it.

File: src/d3d9/d3d9_device.cpp [hunk 3]
```diff
@@ -936,6 +936,9 @@ namespace dxvk {
     const VkImageSubresource dstSubresource = dstTextureInfo->GetSubresourceFromIndex(dstFormatInfo->aspectMask, dst->GetSubresource());
     const VkImageSubresource srcSubresource = srcTextureInfo->GetSubresourceFromIndex(srcFormatInfo->aspectMask, src->GetSubresource());
 
+    if (unlikely((srcSubresource.aspectMask & (VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT)) && m_activeBeginScene))
```
Comment (K0bin) [inline src/d3d9/d3d9_device.cpp:939]: I'm not quite up to date, does StretchRect work on DS textures in DXVK? If not, you should add a TODO comment/warning while you're at it.

File: src/d3d9/d3d9_device.h [hunk 4]
```diff
@@ -1075,6 +1075,8 @@ namespace dxvk {
     D3D9ConstantLayout              m_vsLayout;
     D3D9ConstantLayout              m_psLayout;
 
+    bool                            m_activeBeginScene = false;
```
[the remark this answers is not in the document]
Comment (DadSchoorse) [reply src/d3d9/d3d9_device.h]: naming things at 1am is hard :frog:

File: src/d3d9/d3d9_device.cpp (hunk 3 above)
Comment (DadSchoorse) [reply src/d3d9/d3d9_device.cpp]: It seems to handle resolve/copy, the blit path is a bit questionable.

File: src/d3d9/d3d9_device.cpp [hunk 5]
```diff
@@ -1359,13 +1359,27 @@ namespace dxvk {
   // Some games don't even call them.
 
   HRESULT STDMETHODCALLTYPE D3D9DeviceEx::BeginScene() {
+    D3D9DeviceLock lock = LockDevice();
+
+    if (unlikely(m_inScene))
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:1367]: Maybe just make this a flag?

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_device.h [hunk 6]
```diff
@@ -1169,7 +1169,7 @@ namespace dxvk {
         const     uint32_t regCountHardware = DetermineHardwareRegCount<ProgramType, ConstantType>();
         constexpr uint32_t regCountSoftware = DetermineSoftwareRegCount<ProgramType, ConstantType>();
 
-        if (StartRegister + Count > regCountSoftware)
+        if (uint64_t(StartRegister) + Count > regCountSoftware)
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.h:1172]: when are we ever going yo have more than 4 billion constants?

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_device.h (hunk 6 above)
Comment (DadSchoorse) [reply src/d3d9/d3d9_device.h:1172]: When the application says we should write/read some there.

Comment (Joshua-Ashton) [reply src/d3d9/d3d9_device.h:1172]: why not just make this regCountSoftware - Count then

Comment (DadSchoorse) [reply src/d3d9/d3d9_device.h:1172]: That could underflow.

**Comment by Joshua-Ashton** [comment]
Looks good aside from that I really don't like `[d3d9] Avoid integer overflow in shader constant range check`. Can we drop that for now?

**Comment by DadSchoorse** [comment]
Removed the commit.

**Pull request merged**